rm(list=ls()) 
# # Install BiocManager package and TCGAbiolink
# if (!requireNamespace("BiocManager", quietly=TRUE))
#  install.packages("BiocManager")
# BiocManager::install("TCGAbiolinks")
# # TCGAbiolinks version 2.22.4
# BiocManager::install("EDASeq")
# BiocManager::install("org.Hs.eg.db")
# install.packages("survminer")
# install.packages('DT')
# install.packages("tidyverse")
# BiocManager::install("SummarizedExperiment")
# install.packages("gtsummary")
# install.packages("broom.helpers")
{
  library(SummarizedExperiment)
  library(DT)
  library(TCGAbiolinks)
  library(tidyverse)
  library(org.Hs.eg.db)
  library(survival)
  library(survminer)
  library(readxl)
  library(gtsummary)
}

# Acquiring TCGA datasets
{
  # prep data from TCGA
  query <- GDCquery(
    project = "TCGA-BRCA", 
    data.category = "Transcriptome Profiling", 
    data.type = "Gene Expression Quantification", 
    experimental.strategy = "RNA-Seq", 
    access = "open"
  )
  GDCdownload(
    query, 
    directory = "survivalanalysis"
  )
  # Prepare expression matrix with geneID in the rows and samples (barcode) in the columns
  data <- GDCprepare(
    query, 
    directory = "survivalanalysis", 
    save = TRUE, 
    save.filename = "survivalanalysis/tcga_brca_data_2.15.RDS"
  )
}

##### NOTE: IMPORT "survivalanalysis/tcga_brca_data_2.15.RDS" TO START FROM HERE!!! (generated by above code!)
##### NOTE: IMPORT "survivalanalysis/tcga_brca_data_2.15.RDS" TO START FROM HERE!!! (generated by above code!)
##### NOTE: IMPORT "survivalanalysis/tcga_brca_data_2.15.RDS" TO START FROM HERE!!! (generated by above code!)
##### NOTE: IMPORT "survivalanalysis/tcga_brca_data_2.15.RDS" TO START FROM HERE!!! (generated by above code!)

# Using TCGA data to find prog genes and generate Kaplan-Meier survival plots
{
  # Creating "data.mtx" and "survivaldf"
  # "data.mtx" = TCGA TPM data for genes and patients
  # "survivaldf" = Patient clinical metadata for generating survival plots
  {
    # create expression matrix with geneID in rows and samples (barcode) as col
    data.mtx <- assay(data, "tpm_unstrand")
    colnames(data.mtx) <- gsub("\\-\\w*\\-\\w*\\-\\w*\\-\\w*$", "", colnames(data.mtx))
    rows <- mapIds(org.Hs.eg.db, gsub('\\.\\d*', '',rownames(data.mtx)), keytype="ENSEMBL", column="SYMBOL", multiVals = "first")
    rows <- unname(rows)
    rownames(data.mtx) <- rows
    ## getting rid of zero rows and rows with name "NA"
    data.mtx <- data.mtx[!is.na(rownames(data.mtx)),]
    data.mtx <- data.mtx[!rowMedians(data.mtx)<=5,]
    
    # get subtype information
    data.subtype <- TCGAquery_subtype(tumor = "BRCA")
    data.subtype <- data.subtype[data.subtype$BRCA_Subtype_PAM50 %in% c('Basal'),] #Basal: TNBC
    ## Upload updated clinical data. Only use the data.subtype if the information is not available 
    ## patient data is found in: Liu, J., et al., Cell 2018
    
    Clinical_data <- read_excel("survivalanalysis/patient_data.xlsx")
    Clinical_data <- Clinical_data[Clinical_data$bcr_patient_barcode %in% (data.subtype$patient),]
    Clinical_data <- Clinical_data %>%
      mutate(OS.time = abs(as.numeric(Clinical_data$OS.time)/365.25))
    
    rm(data.subtype)
    
    survivaldf <- data.frame(patient = Clinical_data$bcr_patient_barcode, 
                             status = Clinical_data$OS,
                             time = Clinical_data$OS.time)
    survivaldf <- survivaldf[survivaldf$patient %in% colnames(data.mtx),]
    write.csv(survivaldf,"survivalanalysis/survivaldf.csv")
  }
  
  # Populate "survival_genes_worse", "survival_genes_better",
  #  and provide TPM high/low bins for each gene to "survivaldf"
  {
    # survival_genes_worse = Prognostic_worse_outcome_genes
    survival_genes_worse <- character()
    # survival_genes_better = Prognostic_better_outcome_genes
    survival_genes_better <- character()
    
    Genes <- rownames(data.mtx)
    for (i in 1:length(Genes)) {
      ## Extracting genes and tpms
      gene <- Genes[i]
      tpms <- data.mtx[i,]
      df <- survivaldf
      df$tpm <- tpms[match(df$patient,names(tpms))]
      ## Determine the optimal cut-point for each gene. Minimum proportion of observation group: 10% (Default).
      cut <- surv_cutpoint(df, time = "time", event = "status", variables = c('tpm'), minprop = 0.1)
      ## Divide each variable values based on the cut-point
      # Labels entries in df (stored in cut) that are high/low based on cutpoint vs tpm
      data.cut <- surv_categorize(cut)
      ## Computes an estimate of a survival curve for censored data using either the Kaplan-Meier
      fit <- survfit(Surv(time, status) ~ tpm, data = data.cut)
      #Calculating HR for each gene:
      df2s <- coxph(Surv(time, status)~tpm=='high', data = data.cut)
      df2s <- df2s$coefficients
      df2s <- exp(df2s)
      ## If the survival pvalue is smaller than 0.05, the gene is considered significant.
      if (surv_pvalue(fit)$pval <= 0.1) {
        # Overwrite existing survivaldf to append high/low tpm data for this gene
        survivaldf[gene] <- data.cut$tpm
        if (surv_pvalue(fit)$pval <= 0.05) {
          #Prognostic significant genes: HR>=1.2 p<=0.05
          if (df2s>=1.2) {
            survival_genes_worse <- append(survival_genes_worse, gene)
          }
          if (df2s<=1/1.2){
            survival_genes_better <- append(survival_genes_better,gene)
          }
        }
      }
      rm(gene, tpms, df, cut, data.cut, fit, df2s)
    }
    write.table(survival_genes_better,"survivalanalysis/Prognostic_better_outcome_genes_0.05q_1.2HR.csv", sep=",", col.names = F, row.names = F, quote = F)
    write.table(survival_genes_worse,"survivalanalysis/Prognostic_worse_outcome_genes_0.05q_1.2HR.csv", sep=",", col.names = F, row.names = F, quote = F)
  }
  
  # Shared function to create all survival plots and associated data
  #  Generates: "survival-curve_<TAG>.pdf", "gene-list_<TAG>.csv", and "survival-curve_<TAG>.csv"
  #   "survival-curve_<TAG>.pdf" = survival curve of patients with a collective gene expression
  #   "gene-list_<TAG>.csv" = list of genes used to create the survival curve
  #   "survival-curve_<TAG>_STATS.csv" = coxph() stats for the "high" curve, including Hazard Ratio
  {
    # Creating blank, global-scope surdf_overlap
    # Could not get coxph() to see surdf_overlap when passed as a local variable of plot_survival_curve()
    surdf_overlap <- c()
    
    ## Survival curve plotting
    plot_survival_curve <- function(
    genes, # Data frame of genes (optionally including statistics from their source analysis)
    out_path, # Location this survival curve + associated data will be generated
    tag, # Common name of the data generated
    time_cut=10, # Lets you truncate the X axis of your KM plot
    palette = c("#f3766e", "#4472c4") # Order is (high, low)
    ){
      # Write out list of genes provided
      write_csv(genes,paste0(out_path, "gene-list_",tag,".csv"))
      
      # "survivaldf" has for each patient (row):
      #   "low/high" entries for every gene in data.mtx
      #   Clinical data of patient barcode, "status" as 0/1, and "time"
      # "survivaldf" only needs to be generated once from TCGA data.
      # Then its gene high/low information can be used to construct KM plots
      # "surdf_overlap" is a subset of "survivaldf" which only includes "gene_names" genes
      gene_names <- genes$gene_name
      surdf_overlap <<- survivaldf[,colnames(survivaldf) %in% c(gene_names,"patient","status","time")]
      
      # Adding column "Row_mode"
      # Row_mode is aggregate gene high/low outcomes based on the mode of a patient's TPMs.
      # Does not account for the statistical relevance of each gene.
      # Gene relevance comes from constraining the gene list beforehand (in our case, pval <= 0.05)
      surdf_overlap$Row_mode<<-apply(
        # Operate on new, filtered data frame from survivaldf
        surdf_overlap,
        # Pass each row to function(x) separately
        1,
        # Aggregates gene tpm data into a combined "high" or "low" signal based on majority (mode) entries
        function(x) {
          # Returns the factor with the highest count for a patient's row in surdf_overlap
          names(
            # Returns the table index of the highest count factor
            which.max(
              # Creates a table of counts for each factor
              table(
                # Patient's entries converted to factors for counting
                factor(
                  x,
                  # Create factor labels based on row data
                  unique(x)
                ))))})
      
      
      # Calling survfit() from survminer
      # https://rpkgs.datanovia.com/survminer/reference/surv_fit.html
      fit <- survfit(Surv(time, status) ~ Row_mode, data = surdf_overlap) 
      
      # Add method to grid.draw so survival plots can be created w/ ggsave()
      # https://github.com/kassambara/survminer/issues/152#issuecomment-938941051
      grid.draw.ggsurvplot <<- function(x){
        survminer:::print.ggsurvplot(x, newpage = FALSE)
      }
      
      # Creating survival curve plot from fit data
      p <- ggsurvplot(fit, pval = F, risk.table = T, 
                      xlim = c(0, time_cut), break.time.by = 1,
                      tables.y.text = F,
                      xlab = "Time (yrs)",
                      title = paste0(tag, " (", length(gene_names)," genes)"),
                      palette = palette
      )
      
      # Automatically creating labels for "high" and "low"
      {
        # Manually choosing dataset in plot w/ most points
        # If you lose these "High" and "Low" labels, this line is your likely culprit
        pdata <- p$plot$layers[[1]]$data
        # Finding last relevant points
        plast_high <- tail(pdata[pdata$time < time_cut & pdata$Row_mode=="high",],n=1)
        plast_low <- tail(pdata[pdata$time < time_cut & pdata$Row_mode=="low",],n=1)
        # Adding text labels for high and low
        p$plot <- p$plot +
          annotate("text",label = paste0("High"), color = palette[1],
                   x = Inf,y = as.numeric(plast_high["surv"])-0.01, hjust = 1, vjust = 1, size=5, fontface=2) +
          annotate("text",label = paste0("Low"), color = palette[2],
                   x = Inf,y = as.numeric(plast_low["surv"])-0.01, hjust = 1, vjust = 1, size=5, fontface=2)
      }
      
      # Creating statistics for survival data
      t <- coxph(Surv(time, status) ~ Row_mode=='high', data = surdf_overlap) %>% tbl_regression(exp=TRUE)
      write.csv(t$table_body, paste0(out_path, "survival-curve_", tag, "_STATS.csv"))
      
      # Display survival statistics on the plot
      p$plot <- p$plot +
        annotate("text",
                 label = paste0(
                   "      'high' stats:",
                   "\n   p (KM) = ",if (surv_pvalue(fit)$pval < 0.0001){
                     format(surv_pvalue(fit)$pval, digits=3, scientific=T)
                   }else{
                     format(surv_pvalue(fit)$pval, digits=3, scientific=F)
                   },
                   "\n   N = ",t$table_body$N[3],
                   "\n   HR = ",round(t$table_body$estimate[3],digits=2),
                   " (",round(t$table_body$conf.low[3], digits=2),"-",round(t$table_body$conf.high[3], digits=2),")",
                   "\n   p (HR) = ",if (t$table_body$p.value[3] < 0.0001){
                     format(t$table_body$p.value[3], digits=3, scientific=T)
                   }else{
                     format(t$table_body$p.value[3], digits=3, scientific=F)
                   },"\n\n"),
                 x = -Inf,y = -Inf, hjust = 0, vjust = 0, size=5, fontface=2)
      
      # Saving plot to a pdf file
      ggsave(filename=paste0(out_path, "survival-curve_", tag, ".pdf"), plot = p, device = "pdf", width = 6, height = 9)
    }
  }
  
  # Creating Kaplan-Meier curves for our DESeq analysis hits (filtering using prognosis genes)
  {
    genes <- read_csv("../2-DESeq-R/results/intersect_PlanA_(10w)U(24h)_BOTH-FC-by-acidosis_0.1FDR.csv")
    out_path = "PlanA/"
    
    # Activated by acidosis
    {
      genes_up <- genes[genes$log2FoldChange.MDA24h10w_24h > 0 &
                          genes$log2FoldChange.MDA24h10w_10w > 0,]
      genes_up_TCGAbetter <- genes_up[genes_up$gene_name %in% survival_genes_better,]
      genes_up_TCGAp0.1 <- genes_up[genes_up$gene_name %in% colnames(survivaldf),]
      write_csv(x = genes_up_TCGAp0.1, file = "UP_TCGAp0.1.csv")
      
      # Running genes
      plot_survival_curve(genes=genes_up, out_path=out_path, tag="planA_BOTHfcUP_NOprog", palette = c("#44546a","#1cbdc3"))
      
      # Running w/ prog filter
      plot_survival_curve(genes=genes_up_TCGAbetter, out_path=out_path, tag="planA_BOTHfcUP_progBETTER", palette = c("#44546a","#1cbdc3"))
    }
    
    # Repressed by acidosis
    {
      genes_down <- genes[genes$log2FoldChange.MDA24h10w_24h < 0 &
                            genes$log2FoldChange.MDA24h10w_10w < 0,]
      genes_down_TCGAworse <- genes_down[genes_down$gene_name %in% survival_genes_worse,]
      genes_down_TCGAp0.1 <- genes_down[genes_down$gene_name %in% colnames(survivaldf),]
      write_csv(x = genes_down_TCGAp0.1, file = "DOWN_TCGAp0.1.csv")
      
      # Running genes
      plot_survival_curve(genes=genes_down, out_path=out_path, tag="planA_BOTHfcDOWN_NOprog", palette = c("#f3766e", "#4472c4"))
      
      # Running w/ prog filter
      plot_survival_curve(genes=genes_down_TCGAworse, out_path=out_path, tag="planA_BOTHfcDOWN_progWORSE", palette = c("#f3766e", "#4472c4"))
    }
  }
}
