# RNA-seq raw-read processing, aligning, and quantifying
## Background
Ultimately, our DESeq analysis only makes use of RNA-seq data from two experiments: MDA-MB-231 10 weeks and MDA-MB-231 24 hours.

**NOTE 1:** _The count matrices provided by this repository includes additional sample sets which were not used to produce figures. We filter these with numbered indices. Those indices require updating if you are remaking our count matrices._

# Pipeline summary

## DDS creation
DDS refers to a DESeq Data Set. We use [DESeq2](https://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) to create one data set from our MDA-MB-231 10 week and 24 hour datasets.

- To do this, we exclude samples which do not meet this criteria: `samples_exclude = c(7:12,21:44)`
- The number indices for `samples_exclude` can be determined by checking the positions in `2-DESeq-R/star_samples/meta.csv`

```R
##### DDS creation
{
  source ('modules/run_DESeq.R')
  
  # MDA-MB-231 24h + 10w
  {
    dds_mda_24h10w_samples <- get_dds(group_name = "mda_24h10w_samples",
        combined_categories = list(
        c("duration","condition")
        ),
    design = ~ combined_1,
    samples_exclude = c(7:12,21:44), # MDA 23d, Corbet 6w, and SUM-159
    test_name = "test_mda_24h",
    test_contrast = c("combined_1", "24h_acidosis", "24h_normal"))
  }
}
```

The function `get_dds()` is defined in `2-DESeq-R/modules/run_DESeq.R` and follows these steps for each call:

1. Load paired and single count matrices from `./star_samples/count_matrix_noENSG` by default and combine them into a single count matrix `counts` using `cbind()`. We also polish the column names using some replacement rules specific to our file names.
```R
# Reading count matrix generated by featureCounts
counts_paired <- read.csv(paste0(counts_filenames,"_paired.tsv"), sep = "\t")
counts_paired <- counts_paired[-c(1:6)]
counts <- read.csv(paste0(counts_filenames,"_single.tsv"), sep = "\t")
rownames(counts) <- counts$Geneid
counts <- counts[-c(1:6)]
counts <- cbind(counts, counts_paired)

# Getting sample names from their file names (this is hard-coded for now, a passsed regex parameter would be better)
colnames(counts) <- colnames(counts) %>% str_replace("...5.sort.", "") %>% str_replace("_sorted.bam", "") %>% str_replace_all("\\.","-")
```

2. Exclude columns from `counts` based on indices from `samples_exclude`.
```R
if (length(samples_exclude) > 0){
counts <- counts[-samples_exclude]
}
```

3. Load metadata from `2-DESeq-R/star_samples/meta.csv` and also exclude rows based on indices from `samples_exclude`.
```R
# Preparing the metadata file by adding a combined field
meta <- read.csv(meta_filename,header=TRUE)
if (length(samples_exclude) > 0){
meta <- meta[-samples_exclude,]
}
rownames(meta) <- meta$X
meta <- meta[-c(1)]
```

4. Combine categories from metadata to specify additional groups. In our case, combining `duration` and `condition` so we can call contrasts as `contrast = c("combined_1", "24h_acidosis", "24h_normal")` to specify duration-specific acidosis treatments.
```R
i<-0
for (combined_category in combined_categories){
i<-i+1
meta[paste0("combined_",i)] <- apply(meta[,combined_category], 1, paste, collapse = "_")
}
```

5. Generate and export the model matrix for troubleshooting / transparency of the process. Not entirely necessary, but makes for a nice visual when explaining some innerworkings and constraints of DESeq.
```R
# Export model matrix
mm <-model.matrix(design, meta)
write.csv(mm, file=paste0("results/",group_name,"/",group_name,"_model-matrix.csv"))
```

6. Creating the DESeq Data Set, trimming low-count rows, running DESeq on the dataset, and exporting the modified count matrix for this run.
```R
# The main dds and design definition
dds <-DESeqDataSetFromMatrix(counts, meta, design)
# refining data
ddstrim <- dds[rowSums(counts(dds)) > 10,]
ddstrim <- DESeq(ddstrim)

write.csv(counts(object = ddstrim, normalized=TRUE), file=paste0("results/",group_name,"/",group_name,"_norm-count-matrix.csv"))
```

7. PCA plots are defined in terms of the DDS.
```R
# transform dds for pca analysis
vsd_ddstrim <- assay(vst(ddstrim))
pca_data <- pca(vsd_ddstrim, metadata = meta, removeVar = 0.1)
```

8. p value histogram before and after demonstrates efficacy of removing low count entries in step 6. To do this, it generates results for one contrast (specified as `test_contrast`).

## Calling results and significance
Here we create two lists that contain results called from the above DDS. `res` contains all results while `sig_res` contains a copy of `res` entries filtered for a false discovery rate of `cutoff_FDR` (for this analysis, `cutoff_FDR = 0.1`).

The function `get_results()` accepts a DDS and produces results using the DESeq2 function `results()` for the `contrast` provided. Before returning results, they have rows with NA entries filtered out and gene names deduplicated.

```R
# Relevant lines in get_results() definition
res <- results(dds, contrast = contrast)
res <- res[complete.cases(res),] #remove any rows with NA

res_named <- as.data.frame(res) %>%
    tibble::rownames_to_column(var = "gene_name") %>%
    arrange(padj, abs(log2FoldChange))
res_named <- res_named[!res_named$gene_name %in% NA,]
res_named <- res_named[!duplicated(res_named$gene_name),]

write.csv(res_named, file=paste0("results/",group_name,"/acidosis-vs-control_anyq_anyFC_",results_name,".csv"))
write.csv(res_named[res_named$padj < cutoff_FDR,], file=paste0("results/",group_name,"/Acidosis-vs-control_q",cutoff_FDR,"_anyFC_",results_name,".csv"))
```

## Matching to prog genes
Prognosis genes are loaded from TCGA analysis. TCGA analysis requires results from the above block to be generated. These genes are compared with DESeq results in later blocks.
```R
# Loading results from new TCGA analysis
prog_BETTER = read.csv('../3-TCGA-R/PlanA/gene-list_planA_BOTHfcUP_progBETTER.csv')
prog_WORSE = read.csv('../3-TCGA-R/PlanA/gene-list_planA_BOTHfcDOWN_progWORSE.csv')
```

## Generating intersects
Intersects are created from different results by matching these `gene_name` entries. We call `directional_intersect()` to automatically generate combinations of sig-FDR, any-FDR, any-FC, and both-FC intersecting genes from result sets and export them. This uses `merge_intersecting_genes()`, which is effectively a wrapper around `merge()` to allow for merging multiple sets of genes if desired. The functions are defined in `2-DESeq-R/modules/intersect.R`.
```R
planA <- directional_intersect(list(
    MDA24h10w_24h=res$mda24h10w_24hAvN,
    MDA24h10w_10w=res$mda24h10w_10wAvN
))
write.csv(planA$sig_BOTH, file = "results/intersect_PlanA_(10w)U(24h)_BOTH-FC-by-acidosis_0.1FDR.csv")
write.csv(planA$sig, file = "results/intersect_PlanA_(10w)U(24h)_ANY-FC-by-acidosis_0.1FDR.csv")
```

## Venn/Euler diagrams
This block creates euler diagrams to visualize proportions of intersecting gene lists using the [eulerr R package](https://cran.r-project.org/web/packages/eulerr/vignettes/introduction.html). The function `create_eulerr` is defined in `2-DESeq-R/modules/euler_diagram.R`.
```R
euler_plot <- create_eulerr(
    sig_res$mda24h10w_24hAvN$gene_name, "Short (MDA 24h)",
    sig_res$mda24h10w_10wAvN$gene_name, "Long (MDA 10w)",
    planA$sig_BOTH$gene_name, "Common BOTH"
)
euler_plot <- grid.arrange(grobs = list(euler_plot), top = "(STAR) Plan A - Overlap of genes altered by acidosis (FDR < 0.1)")
ggsave(euler_plot, filename="results/euler_diagram_planA.pdf", width=6, height=5)
```

## Comparing results with l2fc vs l2fc plots
This block creates a plot of the log2 fold change of the results of MDA-MB-231 10 weeks versus 24 hours. This visualizes similarity in the hits between these experiments. If we observe significant genes having the same direction in l2fc between the two durations of exposure, we can suggest that the two durations affect similar genes in similar ways. The function `create_compare_plot()` is defined in `2-DESeq-R/modules/l2fc.R`.
```R
create_compare_plot(set1 = res$mda24h10w_24hAvN,
    set2 = res$mda24h10w_10wAvN,
    name1 = "(pHe 6.4 / 7.4) MDA-MB-231 24 h",
    name2 = "(pHe 6.4 / 7.4) MDA-MB-231 10 w",
    padj_cutoff = cutoff_FDR,
    out_path = "results/l2fc_planA_MDA24h10w.pdf")
```

## Selective Heatmaps
This block creates heat maps of genes based on their l2fc values. In this case, we generate heat maps for (l2fc UP) intersecting (HIPPO pathway genes) intersecting (prognosis BETTER) and a companion plot for DOWN, ILKmany, and WORSE. Here, "ILKmany" refers to a combination of the pathway lists `c(LA_ILK,LA_cc,LA_int,LA_PI3K,LA_glyc)`. Most of the code in `get_hm()` is for manipulating ranges and color paletes for making the plot. The key parts to note here are:
1. We accept prognosis genes which can be optionally associated with a set of genes.
2. We can generate a plot with a limited number of entries (`TOP_GENES`, `TOP_GENES_HIGH`, and `TOP_GENES_LOW`).
3. We can "fix" rows to the plot to force them to display despite limits.
4. Both a TOP set and ALL set of genes are generated as a heat map and a set of genes required to make the specific plot (`heatmap_hits...csv` and `heatmapEXTENDED...csv` reflect TOP and ALL respectively).
5. We feed results to `get_hm()` with or without filtering for prognosis genes.
6. The `get_hm()` function further calls `plot_heatmap()` defined in `2-DESeq-R/modules/heatmaps.R`.
    - Much like `get_hm()`, most of `plot_heatmap()`'s code is dedicated to preparing it for displaying via the `pheatmap()` function from the [Pretty Heatmaps R package](https://cran.r-project.org/web/packages/pheatmap/pheatmap.pdf).
```R
fullsource <- planA$intersect
tag <- "24h-U-10w"

hmsource <- planA$sig_UP[c(1,3,9)]
colnames(hmsource) <- c("gene_name", "MDA 24h", "MDA 10w")
get_hm(pathways="HIPPO", tag=tag, FIXED_ROWS=FIXED_ROWS_HIPPO)
get_hm(pathways="HIPPO",prog="UP", tag=tag, FIXED_ROWS=FIXED_ROWS_HIPPO)

hmsource <- planA$sig_DOWN[c(1,3,9)]
colnames(hmsource) <- c("gene_name", "MDA 24h", "MDA 10w")
get_hm(pathways="ILKmany", tag=tag, FIXED_ROWS=FIXED_ROWS_ILK)
get_hm(pathways="ILKmany",prog="DOWN", tag=tag, FIXED_ROWS=FIXED_ROWS_ILK)
```

## Pathway Analysis
Here we visualize the results of [QIAGEN's Ingenuity Pathway Analysis](https://digitalinsights.qiagen.com/products-overview/discovery-insights-portfolio/analysis-and-visualization/qiagen-ipa/) program. The genes fed to IPA in this case come from "PlanA" referring to the BOTH-FC intersection of MDA-MB-231 10 weeks and 24 hours. The script defining `toppathwaysIPA()` is located at `2-DESeq-R/modules/pathways.R`. We use an added column `combined` which is the product of `-log(pvalue)` and `z-score` to rate pathways by significance and "direction". The function `toppathwaysIPA()` also allows for exclusion of pathways and to set a limit on the number of displayed pathways. In our case, we use exclusions to remove redundant/subtype pathways already displayed.
```R
  # "(MDA 24h) U (MDA 10w)"
  # IPA_PATHWAYS_STAR_MDA24h_vs10W.csv
  toppathwaysIPA(input = "GO/IPA_PATHWAYS_STAR_MDA24h_vs10W.csv", tag = "((MDA 24h) U (MDA 10w)) BOTH Common", limit = 50, exclude_from_top_by_index = c())
  toppathwaysIPA(input = "GO/IPA_PATHWAYS_STAR_MDA24h_vs10W.csv", tag = "((MDA 24h) U (MDA 10w)) BOTH Common REDUCED", limit = 33, exclude_from_top_by_index = c(3,5,6,11,12,19,20,23:26,28:30,42,46,48),height=7)
  toppathwaysIPA(input = "GO/IPA_PATHWAYS_STAR_MDA24h_vs10W.csv", tag = "((MDA 24h) U (MDA 10w)) BOTH Common TOP20plus", limit = 26, exclude_from_top_by_index = c(21:26,28:32,34:44,46,48),height=6)
  toppathwaysIPA(input = "GO/IPA_PATHWAYS_STAR_MDA24h_vs10W.csv", tag = "((MDA 24h) U (MDA 10w)) BOTH Common TOP10plus", limit = 17, exclude_from_top_by_index = c(11:13,15:26,28:32,34:44,46,48),height=6)
```

# Terminology in file names:
- AvN = Acidosis vs. Normal (DESeq contrast)
- q0.1 = Filtered for a false discovery rate (padj, qval, FDR) less than 0.1
- ANY-FC = Not filtered by log2FoldChange.
- BOTH-FC = Significantly differentially expressed genes "Common" in both Long and Short results with same log2FoldChange direction. `((Long DOWN) ∪ (Short DOWN)) + ((Long UP) ∪ (Short UP))`

- ILK = Integrin linked kinase (gene pathways aggregated from multiple sources)
- CC = Cell cycle (")
- INT = Integrin (")
- PI3K = PI3K/AKT/mTOR pathway (")
- GLYC = Glycolysis (")
- HIPPO = Hippo signaling pathway (")
_Lists of genes assigned to these pathways can be provided upon request, and will be present in source code_

- prog_BETTER = Genes associated with a BETTER OS from the TCGA survival analysis
- prog_WORSE = Genes associated with a WORSE OS from the TCGA survival analysis